config/store.go

```diff
@@ -188,124 +196,155 @@ func (s *Store) Set(newCfg *model.Config) (*model.Config, error) {
 		return nil, errors.Wrap(err, "new configuration is invalid")
 	}
 
-	newCfg = removeEnvOverrides(newCfg, s.configNoEnv, s.GetEnvironmentOverrides())
+	newCfgNoEnv := removeEnvOverrides(newCfg, oldCfgNoEnv, s.GetEnvironmentOverrides())
 
-	// Don't persist feature flags unless we are on MM cloud
+	// Don't store feature flags unless we are on MM cloud
 	// MM cloud uses config in the DB as a cache of the feature flag
 	// settings in case the management system is down when a pod starts.
-	if !s.persistFeatureFlags {
+
+	// Backing up feature flags section in case we need to restore them later on.
+	oldCfgFF := oldCfg.FeatureFlags
+	oldCfgNoEnvFF := oldCfgNoEnv.FeatureFlags
+	// Clearing FF sections to avoid both comparing and persisting them.
+	if s.readOnlyFF {
+		oldCfg.FeatureFlags = nil
 		newCfg.FeatureFlags = nil
+		newCfgNoEnv.FeatureFlags = nil
 	}
 
-	if err := s.backingStore.Set(newCfg); err != nil {
+	if err := s.backingStore.Set(newCfgNoEnv); err != nil {
 		return nil, errors.Wrap(err, "failed to persist")
 	}
 
-	if err := s.loadLockedWithOld(oldCfg, &unlockOnce); err != nil {
-		return nil, errors.Wrap(err, "failed to load on save")
+	newCfg = applyEnvironmentMap(newCfgNoEnv, GetEnvironment())
```

Comment:
Looking deeper into this, `s.GetEnvironmentOverrides` work on the same set as `GetEnvironment` as the previous is just:

```
func (s *Store) GetEnvironmentOverrides() map[string]interface{} {
	return generateEnvironmentMap(GetEnvironment(), nil)
}
```
So is there a benefit in remove env overrides, and then adding the same overrides again? Can't we just reuse `newCfg`?

Reply:
Doing that would be assuming `newCfg` (user input) has the environment overrides already set. There are cases where it might not be like that hence the need to remove/apply them anyway.

Reply:
Ok, perhaps makes sense to add a comment to clarify that? It gets hard to keep all of this in mind.